File: midonet-cluster/src/main/java/org/midonet/cluster/auth/keystone/v2_0/KeystoneClient.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.cluster.auth.keystone.v2_0;

import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.jaxrs.json.JacksonJaxbJsonProvider;
import com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider;
import com.sun.jersey.api.client.Client;
import com.sun.jersey.api.client.ClientHandlerException;
import com.sun.jersey.api.client.UniformInterfaceException;
import com.sun.jersey.api.client.WebResource;
import com.sun.jersey.api.client.config.DefaultClientConfig;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.midonet.cluster.auth.KeystoneAccess;
import org.midonet.cluster.auth.KeystoneAuthCredentials;
import org.midonet.cluster.auth.KeystoneTenant;
import org.midonet.cluster.auth.KeystoneTenantList;
import org.midonet.cluster.package$;

/**
 * Keystone Client V2.0
 */
public class KeystoneClient {

    private final static Logger log = LoggerFactory
            .getLogger(package$.MODULE$.keystoneLog());

    private final KeystoneConfig config;

    public static final String KEYSTONE_TOKEN_HEADER_KEY = "X-Auth-Token";
    public static final String MARKER_QUERY = "marker";
    public static final String LIMIT_QUERY = "limit";

    private final DefaultClientConfig clientConfig = new DefaultClientConfig();

    public KeystoneClient(KeystoneConfig config) {
        this.config = config;

        JacksonJsonProvider jsonProvider = new JacksonJaxbJsonProvider();
        jsonProvider.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES,
                               false);

        this.clientConfig.getSingletons().add(jsonProvider);
    }

    public String getServiceUrl() {
        return config.protocol() + "://" + config.host() + ":" +
               Integer.toString(config.port()) + "/v2.0";
    }

    private String getTokensUrl() {
        return getServiceUrl() + "/tokens";
    }

    private String getTokensUrl(String token) {
        return getTokensUrl() + "/" + token;
    }

    private String getTenantsUrl() {
        return getServiceUrl() + "/tenants";
    }

    private String getTenantUrl(String id) {
        return getTenantsUrl() + "/" + id;
    }

    private <T> T sendGetRequest(WebResource resource, Class<T> clazz)
            throws KeystoneServerException, KeystoneConnectionException {

        try {
            return resource.accept(MediaType.APPLICATION_JSON)
                    .header(KEYSTONE_TOKEN_HEADER_KEY, config.adminToken())
                    .get(clazz);
        } catch (UniformInterfaceException e) {
            throw new KeystoneServerException("Keystone server error.", e,
                    e.getResponse().getStatus());
        } catch (ClientHandlerException e) {
            throw new KeystoneConnectionException(
                    "Could not connect to Keystone server " + resource.getURI(),
                    e);
        }
    }


    public KeystoneAccess createToken(KeystoneAuthCredentials credentials)
            throws KeystoneServerException, KeystoneConnectionException,
                   KeystoneUnauthorizedException {

        Client client = Client.create(clientConfig);
        WebResource resource = client.resource(getTokensUrl());

        try {
            return resource.type(MediaType.APPLICATION_JSON)
                    .accept(MediaType.APPLICATION_JSON)
                    .post(KeystoneAccess.class, credentials);
        } catch (UniformInterfaceException e) {
            if (e.getResponse().getStatus()
                == Response.Status.BAD_REQUEST.getStatusCode()) {
                String err = "Keystone v2 login credentials invalid " +
                             credentials.toString();
                log.warn(err);
                throw new KeystoneUnauthorizedException(err, e);
            } else if (e.getResponse().getStatus()
                == Response.Status.UNAUTHORIZED.getStatusCode()) {
                String err = "Keystone v2 login credentials not found " +
                             credentials.toString();
                log.warn(err,credentials);
                throw new KeystoneUnauthorizedException(err, e);
            } else {
                throw new KeystoneServerException("Keystone server error", e,
                    e.getResponse().getStatus());
            }
        } catch (ClientHandlerException e) {
            throw new KeystoneConnectionException(
                    "Could not connect to Keystone server. Uri="
                            + resource.getURI(), e);
        }
    }

    public KeystoneAccess getToken(String token) throws KeystoneServerException,
            KeystoneConnectionException {

        Client client = Client.create(clientConfig);
        WebResource resource = client.resource(getTokensUrl(token));

        try {
            return sendGetRequest(resource, KeystoneAccess.class);
        } catch (KeystoneServerException ex) {
            if (ex.getStatus() == Response.Status.NOT_FOUND
                    .getStatusCode()) {
                // This indicates that the token was not found
                log.warn("Keystone v2 token not found {}", token);
                return null;
            }

            throw ex;
        }
    }

    /**
     * Retrieves from Keystone server a tenant given its ID.
     *
     * @param id  ID of the tenant
     * @return {@link KeystoneTenant} object
     * @throws KeystoneServerException
     * @throws KeystoneConnectionException
     */
    public KeystoneTenant getTenant(String id) throws KeystoneServerException,
            KeystoneConnectionException {
        log.debug("Keystone v2 get tenant {}", id);

        Client client = Client.create(clientConfig);
        WebResource resource = client.resource(getTenantUrl(id));

        try {
            return sendGetRequest(resource, KeystoneTenant.class);
        } catch (KeystoneServerException ex) {
            if (ex.getStatus() == Response.Status.NOT_FOUND.getStatusCode()) {
                // This indicates that the tenant was not found.  Not an error
                // from MidoNet.
                log.info("Keystone v2 tenant not found {}", id);
                return null;
            }

            throw ex;
        }
    }

    /**
     * Retrieves from Keystone server a list of tenants from the given optional
     * marker and limit values.  The marker and limit values are appended to
     * the actual HTTP request to Keystone as query string parameters.  The
     * return object of {@link KeystoneTenantList} represents the response body
     * from the Keystone API containing the tenant list.
     *
     * @param marker The ID of the tenant fetched in the previous request.
     *               The returned list will start from the tenant after the
     *               tenant with this ID.  A null and empty values are ignored.
     * @param limit  The number of tenants to fetch.  Must be greater than 0.
     *               This field is ignored if the value is less than 1.
     * @return {@link KeystoneTenantList} object, representing the response from
     *         the Keystone API.
     * @throws KeystoneServerException
     * @throws KeystoneConnectionException
     */
    public KeystoneTenantList getTenants(String marker, int limit)
            throws KeystoneServerException, KeystoneConnectionException {
        log.debug("KeystoneClient.getTenants: entered marker=" + marker +
                ", limit=" + limit);

        Client client = Client.create(clientConfig);
        WebResource resource = client.resource(getTenantsUrl());

        if (!StringUtils.isEmpty(marker)) {
            resource = resource.queryParam(MARKER_QUERY, marker);
        }

        if (limit > 0) {
            resource = resource.queryParam(LIMIT_QUERY,
                    Integer.toString(limit));
        }

        return sendGetRequest(resource, KeystoneTenantList.class);
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/auth/keystone/v2/KeystoneClient.scala
/*
 * Copyright 2016 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.auth.keystone.v2

import java.net.URI
import java.time.{ZoneOffset, LocalDateTime, ZonedDateTime}
import java.time.format.{DateTimeParseException, DateTimeFormatter}
import java.util.logging.Logger

import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response.Status
import javax.ws.rs.core.UriBuilder

import scala.concurrent.duration._
import scala.util.Try

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.jaxrs.json.JacksonJaxbJsonProvider
import com.sun.jersey.api.client.config.DefaultClientConfig
import com.sun.jersey.api.client.filter.LoggingFilter
import com.sun.jersey.api.client.{Client, ClientHandlerException, UniformInterfaceException}

import org.apache.commons.lang3.StringUtils.{isBlank, isNotBlank}

import org.midonet.cluster.auth.keystone.v2.KeystoneClient._

object KeystoneClient {

    /**
      * This specifies the interval before the expiration time of the
      * administrative token when it will no longer be used. This ensures that
      * clock difference between the client and the Keystone server does not
      * cause the client to use an expired token.
      */
    private val TokenExpirationGuard = 1 minute

    /**
      * Contains information about an administrative token.
      */
    private case class AdminToken(id: String,
                                  serverIssueTime: Long,
                                  clientIssueTime: Long,
                                  expirationTime: Long)

    /**
      * Parses the given timestamp from the two supported date formats: ISO8601
      * and JDBC to a Unix timestamp. If any parsing fails, the method returns
      * zero.
      */
    @throws[KeystoneException]
    private[keystone] def parseTimestamp(string: String): Long = {
        val tryZonedTime = Try {
            ZonedDateTime.parse(string, DateTimeFormatter.ISO_DATE_TIME)
                         .toInstant.toEpochMilli
        }
        val tryLocalTime = Try {
            LocalDateTime.parse(string, DateTimeFormatter.ISO_DATE_TIME)
                         .toInstant(ZoneOffset.of("Z")).toEpochMilli
        }

        tryZonedTime recoverWith {
            case _ => tryLocalTime
        } getOrElse {
            throw new KeystoneException(
                null, s"Unknown token timestamp format for $string", null)
        }
    }
}


class KeystoneClient(config: KeystoneConfig) {

    private val jsonProvider = new JacksonJaxbJsonProvider()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    private val clientConfig = new DefaultClientConfig()
    clientConfig.getSingletons.add(jsonProvider)

    private val client = Client.create(clientConfig)
    client.addFilter(new LoggingFilter(
        Logger.getLogger("org.midonet.cluster.auth.keystone")))

    @volatile private var adminToken =
        if (isNotBlank(config.adminToken))
            AdminToken(config.adminToken, 0L, 0L, Long.MaxValue)
        else null

    /**
      * Gets the current Keystone version.
      */
    @throws[KeystoneException]
    def getVersion: KeystoneVersion = {
        get(classOf[KeystoneVersion], None)()
    }

    /**
      * Authenticates against Keystone for the given tenant using the specified
      * password credentials. The method returns a [[KeystoneAccess]] instance
      * containing an authorization token for the given user.
      */
    @throws[KeystoneException]
    def authenticate(tenantName: String, userName: String, password: String)
    : KeystoneAccess = {
        post(KeystoneAuth(PasswordAuth(PasswordCredentials(userName, password),
                                      tenantName)),
             classOf[KeystoneAccess], None, "tokens")()
    }

    /**
      * Authenticates against Keystone for the given tenant using the specified
      * token. The method returns a [[KeystoneAccess]] instance containing an
      * authorization token for the user of the given token.
      */
    @throws[KeystoneException]
    def authenticate(tenantName: String, token: String): KeystoneAccess = {
        post(KeystoneAuth(TokenAuth(Token(id = token), tenantName)),
             classOf[KeystoneAccess], None, "tokens")()
    }

    /**
      * Validates an authorization token using the [[KeystoneClient]]'s
      * administrative credentials. If the `tenantScope` argument is true, then
      * the method validates the token in the specified tenant scope for
      * performance.
      */
    @throws[KeystoneException]
    def validate(token: String, tenantScope: Boolean = false): KeystoneAccess = {
        var params = Seq.empty[(String, String)]
        if (tenantScope) params = params :+ ("belongsTo", "")
        withAdminToken { adminToken =>
            get(classOf[KeystoneAccess], Some(adminToken), "tokens", token)(
                params: _*)
        }
    }

    /**
      * Lists the users for the given client token. The client must be an
      * administrator.
      */
    @throws[KeystoneException]
    def listUsers(token: String): KeystoneUsers = {
        get(classOf[KeystoneUsers], Some(token), "users")()
    }

    /**
      * Gets the information for the user with the specified name. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getUserByName(name: String, token: String): KeystoneUser = {
        get(classOf[KeystoneUser], Some(token), "users")(("name", name))
    }

    /**
      * Gets the information for the user with the specified identifier. The
      * client must be an administrator.
      */
    @throws[KeystoneException]
    def getUserById(id: String, token: String): KeystoneUser = {
        get(classOf[KeystoneUser], Some(token), "users", id)()
    }

    /**
      * Lists the tenants for the given client token. The client must be an
      * administrator.
      * @param marker Specifies the identifier of the last seen tenant when using
      *               multi-page listing.
      * @param limit Specifies a limit on the number of items returned by
      *              the operation when using multi-page listing.
      */
    @throws[KeystoneException]
    def listTenants(token: String, marker: Option[String] = None,
                    limit: Option[Int] = None)
    : KeystoneTenants = {
        var params = Seq.empty[(String, String)]
        if (marker.isDefined) params = params :+ ("marker", marker.get)
        if (limit.isDefined) params = params :+ ("limit", limit.get.toString)
        get(classOf[KeystoneTenants], Some(token), "tenants")(params: _*)
    }

    /**
      * Gets the information for the tenant with the specified name. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getTenantByName(name: String, token: String): KeystoneTenant = {
        get(classOf[KeystoneTenant], Some(token), "tenants")(("name", name))
    }

    /**
      * Gets the information for the tenant with the specified identifier. The
      * client must be an administrator.
      */
    @throws[KeystoneException]
    def getTenantById(id: String, token: String): KeystoneTenant = {
        get(classOf[KeystoneTenant], Some(token), "tenants", id)()
    }

    /**
      * Gets the list of roles for the user on the specified tenant. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getTenantUserRoles(tenantId: String, userId: String, token: String,
                           marker: Option[String] = None,
                           limit: Option[Int] = None)
    : KeystoneRoles = {
        var params = Seq.empty[(String, String)]
        if (marker.isDefined) params = params :+ ("marker", marker.get)
        if (limit.isDefined) params = params :+ ("limit", limit.get.toString)
        get(classOf[KeystoneRoles], Some(token), "tenants", tenantId,
            "users", userId, "roles")(params: _*)
    }

    /**
      * Gets the Keystone URI for the given path segments and query
      * parameters.
      */
    private def uriFor(paths: String*)(params: (String, String)*): URI = {
        val builder = UriBuilder
            .fromPath(s"${config.protocol}://${config.host}:${config.port}/v2.0")
            .segment(paths: _*)
        for ((param, value) <- params)
            builder.queryParam(param, value)
        builder.build()
    }

    /**
      * A GET HTTP request to the Keystone server.
      */
    private def get[T](clazz: Class[T], token: Option[String], paths: String*)
                      (params: (String, String)*): T = {
        val uri = uriFor(paths: _*)(params: _*)
        tryRequest(uri) {
            val request = client.resource(uri)
                                .`type`(APPLICATION_JSON)
                                .accept(APPLICATION_JSON)
            if (token.isDefined) {
                request.header("X-Auth-Token", token.get)
            }
            request.get(clazz)
        }
    }

    /**
      * A POST HTTP request to the Keystone server.
      */
    private def post[T, R](obj: T, clazz: Class[R], token: Option[String],
                           paths: String*)
                          (params: (String, String)*): R = {
        val uri = uriFor(paths: _*)(params: _*)
        tryRequest(uri) {
            val request = client.resource(uri)
                                .`type`(APPLICATION_JSON)
                                .accept(APPLICATION_JSON)
            if (token.isDefined) {
                request.header("X-Auth-Token", token)
            }
            request.post(clazz, obj)
        }
    }

    /**
      * Provides exception handling for the given function.
      */
    private def tryRequest[R](uri: URI)(f: => R): R = {
        try {
            f
        } catch {
            case e: UniformInterfaceException
                if e.getResponse.getStatus == Status.UNAUTHORIZED.getStatusCode =>
                throw new KeystoneUnauthorizedException(uri.toString, e)
            case e: UniformInterfaceException =>
                throw new KeystoneException(uri.toString, e)
            case e: ClientHandlerException =>
                throw new KeystoneConnectionException(uri.toString, e)
        }
    }

    /**
      * Calls a function with a current administrative token.
      */
    private def withAdminToken[R](f: (String) => R): R = {
        // If there exists a current administrative token.
        var currentToken = adminToken
        // If there is no token or if the token has or is about to expire, renew
        // the token using the password credentials.
        if ((currentToken eq null) ||
            System.currentTimeMillis() + TokenExpirationGuard.toMillis +
            currentToken.serverIssueTime - currentToken.clientIssueTime >=
            currentToken.expirationTime) {

            if (isBlank(config.userName) || isBlank(config.password) ||
                isBlank(config.tenantName)) {
                throw new KeystoneException(null,
                    "Keystone configuration must provide an administrative " +
                    "tenant, user name and password", null)
            }

            val keystoneAccess = authenticate(config.tenantName,
                                              config.userName,
                                              config.password)
            val token = keystoneAccess.access.token
            currentToken = AdminToken(token.id,
                                      parseTimestamp(token.issuedAt),
                                      System.currentTimeMillis(),
                                      parseTimestamp(token.expires))
            adminToken = currentToken
        }
        f(currentToken.id)
    }

}
